strawberryfields/apps/graph/dense.py

```diff
@@ -0,0 +1,193 @@
+    with local searching. The heuristic method can be set with the ``options`` argument. Methods
+    can contain stochastic elements, where randomness can come from distributions including GBS.
+
+    All elements of the heuristic can be controlled with the ``options`` argument, which should be a
+    dict-of-dicts where the first level specifies the option type as a string-based key,
+    with corresponding value being a dictionary of options for that type. The option types are:
+
+    - ``"heuristic"``: specifying options used by optimization heuristic; corresponding
+      dictionary of options explained further :ref:`below <heuristic>`
+    - ``"backend"``: specifying options used by backend quantum samplers; corresponding
+      dictionary of options explained further in :mod:`~strawberryfields.apps.sample`
+    - ``"resize"``: specifying options used by resizing method; corresponding dictionary of
+      options explained further in :mod:`~strawberryfields.apps.graph.resize`
+    - ``"sample"``: specifying options used in sampling; corresponding dictionary of options
+      explained further in :mod:`~strawberryfields.apps.graph.sample`
+
+    If unspecified, a default set of options is adopted for a given option type.
+
+    .. _heuristic:
+
+    The options dictionary corresponding to ``"heuristic"`` can contain any of the following:
+
+    .. glossary::
+
+        key: ``"method"``, value: *str* or *callable*
+            Value can be either a string selecting from a range of available methods or a
+            customized callable function. Options include:
+
+            - ``"random-search"``: a simple random search algorithm where many subgraphs are
+              selected and the densest one is chosen (default)
+            - *callable*: accepting ``(graph: nx.Graph, nodes: int, iterations: int, options:
+              dict)`` as arguments and returning ``Tuple[float, list]`` corresponding to the
+              density and list of nodes of the densest subgraph found, see :func:`random_search`
+              for an example
+
+    Args:
+        graph (graph_type): the input graph
+        nodes (int): the size of desired dense subgraph
+        iterations (int): number of iterations to use in algorithm
+        options (dict[str, dict[str, Any]]): dict-of-dicts specifying options in different parts
+            of heuristic search algorithm; defaults to :const:`OPTIONS_DEFAULTS`
+
+    Returns:
+        tuple[float, list]: the density and list of nodes corresponding to the densest subgraph
+        found
+    """
+    options = {**OPTIONS_DEFAULTS, **(options or {})}
+
+    method = options["heuristic"]["method"]
+
+    if not callable(method):
+        method = METHOD_DICT[method]
+
+    return method(
+        graph=utils.validate_graph(graph),
+        nodes=nodes,
+        iterations=iterations,
+        options=options,
+    )
+
+
+def random_search(
```

Comment:
This might have been discussed before, but why do we want to include random search as an algorithm? It's a terrible way to find dense subgraphs. Is it to compare with GBS?

Reply:
The idea is to provide a couple of methods, one of them being the most elementary random search. There is also all the functionality for the charikar algorithm in the other modules, it just got missed to put the final functionality into the `dense` module as we switched to max clique, but we can definitely do that time permitting.

Reply:
I would prefer not to even have this as an option. Instead, we just run Charikar as the default algorithm, e.g., apps.find_dense(graph, k)

Reply:
My current line of thinking was to get rid of the `find_dense` wrapper function and make available both `random_search` and `charikar` functions, but we could emphasise `charikar` more in the documentation. By default algorithm, you mean that we'd _only_ offer Charikar? I'm fine with that, although I could imagine random search might have some value to people who read the original paper using random search.